ldap/src/main/java/org/springframework/security/ldap/authentication/ad/ActiveDirectoryLdapAuthenticationProvider.java

```diff
@@ -187,29 +186,28 @@ protected DirContextOperations doAuthentication(
 	}
 
 	private DirContext bindAsUser(String username, String password) {
-		// TODO. add DNS lookup based on domain
-		final String bindUrl = url;
-
-		Hashtable<String, String> env = new Hashtable<>();
-		env.put(Context.SECURITY_AUTHENTICATION, "simple");
 		String bindPrincipal = createBindPrincipal(username);
-		env.put(Context.SECURITY_PRINCIPAL, bindPrincipal);
-		env.put(Context.PROVIDER_URL, bindUrl);
-		env.put(Context.SECURITY_CREDENTIALS, password);
-		env.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory");
-		env.put(Context.OBJECT_FACTORIES, DefaultDirObjectFactory.class.getName());
+
+		// TODO. add DNS lookup based on domain
+		contextSource.setUrl(url);
+		contextSource.setUserDn(bindPrincipal);
```

Comment:
This is a race condition because we are setting a stack variable onto a member variable.